### PR TITLE
Refactor unit tests

### PR DESCRIPTION
**Pull request by mgcrea**
##### Checklist

- [X] tests and code linting passes
- [X] the commit message follows commit guidelines

##### Affected core subsystem(s)

Tests

##### Description of change

- Updated layout for cleaner/easier testing (prepare for future `fixtures` and testing `.eslintrc`).
- Test artefacts are now redirected to `.tmp/test` folder.
- Now using a `mocha.opts` to define mocha options
- Using bdd reporting to get a better testing feedback and performance tracking (with slow tests).

### CODE REVIEW

**Comment by mgcrea** [comment]
Just added a commit refactoring eslint leveraging Airbnb's config which is a pretty good (strict) starting point. Tried to minimize the changes to the code by adding according rules.

One thing that we should definitely change in my opinion is the indenting. The whole JS world uses spaces (both airbnb and feross/standard styleguide do recommend spaces), as an open-source library, this project should try to closely follow commonly used standards.

I did not want to push the change in this commit (should be a single separated commit) without polling you about it as it is often a "sensitive" subject since changing habits is hard. I won't mind if you do not want to switch.

File: package.json [hunk 1]
```diff
@@ -36,15 +36,21 @@
     "babel-register": "^6.9.0",
     "esdoc": "^0.4.7",
     "esdoc-es7-plugin": "0.0.3",
-    "eslint": "^2.8.0",
+    "eslint": "^2.13.1",
+    "eslint-config-airbnb-base": "^3.0.1",
+    "eslint-plugin-import": "^1.9.2",
     "mocha": "^2.3.3",
+    "nyc": "^6.6.1",
```
Comment (alexanderGugel) [inline package.json]: Good idea

File: Makefile [hunk 2]
```diff
@@ -44,16 +43,16 @@ clean:
 	rm -rf lib test/test
 
 lint:
-	$(DEPS_BIN_DIR)/eslint src
+	npm run lint
```
Comment (alexanderGugel) [inline Makefile]: If we simply run the `npm` scripts from the `Makefile`, we should be consistent about it. So I think we should do the same for the `dev` and `watch` targets.

Comment (alexanderGugel) [inline Makefile]: Quite ironic that we're using npm here actually. Why not use `./lib/cmd.js run lint`?

File: src/cache.js [hunk 3]
```diff
@@ -33,7 +33,7 @@ export function getTmp () {
  * @return {WriteStream} - Write Stream
  */
 export function write () {
-	return fs.WriteStream(getTmp())
+	return fs.createWriteStream(getTmp())
```
Comment (alexanderGugel) [inline src/cache.js]: What's the idea behind this change (https://github.com/nodejs/node/blob/master/lib/fs.js#L1827).

File: src/config_cmd.js [hunk 4]
```diff
@@ -6,14 +6,12 @@ import Table from 'easy-table'
  */
 export default function configCmd () {
 	const table = new Table()
-	const keys = Object.keys(config)
-
-	for (let key of keys) {
+	Object.keys(config).forEach(key => {
```
Comment (alexanderGugel) [inline src/config_cmd.js]: Hmmm... more like a matter of taste here really..

File: src/debuglog.js [hunk 5]
```diff
@@ -42,19 +42,19 @@ function debuglog (set) {
 	if (debugEnv === undefined) {
 		debugEnv = process.env.NODE_DEBUG || ''
 	}
-	set = set.toUpperCase()
-	if (!debugs[set]) {
-		if (debugEnv === '*' || new RegExp('\\b' + set + '\\b', 'i').test(debugEnv)) {
+	const upperSet = set.toUpperCase()
```
Comment (alexanderGugel) [inline src/debuglog.js]: I would rather not change this method too much. It's taken directly from Node (https://github.com/nodejs/node/blob/master/lib/util.js#L145).

**Comment by alexanderGugel** [comment]
I'm mostly ok with this. I think there are a lot of good changes in there, but a lot of things are just a matter of taste I think.

E.g. I don't think using `forEach` instead of `for of` is necessarily "better", so I don't really think it makes sense to change that kind of stuff. I don't feel strongly about balanced spacing in params. Not sure if it really makes sense to change it.

I really like the tooling changes, e.g. the Mocha config etc. Also consistently using template strings is a great idea. And we definitely needed test coverage.

Concerning tabs vs spaces: I'm using a tabstop of 8, but I realise that the rest of the world uses either 2 or 4. I think using a tabstop of 8 makes it more painful to write deeply nested functions. I think that's generally speaking not a bad thing.

Any chance we can keep 5702f15 separate and discuss the changes to the linting rules first?

File: Makefile (hunk 2 above)
Comment (mgcrea) [inline Makefile]: 👍 for `./lib/cmd.js`

File: src/cache.js (hunk 3 above)
Comment (mgcrea) [inline src/cache.js]: Were encountering the eslint [new-cap](http://eslint.org/docs/rules/new-cap) rule: `A function with a name starting with an uppercase letter should only be used as a constructor.`

File: src/debuglog.js (hunk 5 above)
Comment (mgcrea) [inline src/debuglog.js]: @alexanderGugel, I'm fine with reverting it, issue was [no-param-reassign](http://eslint.org/docs/rules/no-param-reassign), but I've added exceptions elsewhere, so I will do that as well.

File: src/config_cmd.js (hunk 4 above)
Comment (mgcrea) [inline src/config_cmd.js]: Indeed, but it's 1LOC+1DEF shorter, but I'm fine reverting it.

**Comment by mgcrea** [comment]
Thanks for the review, indeed some changes are cosmetic, and a matter of state, I don't really mind about any of them, a few side-notes:

- regarding `forEach` instead of `for of`, I'm not encountering `for of` in most of the code I'm looking into so I'm not "used to it", `forEach` is shorter of both 1 LOC and 1 variable definition so it looks like a little bit more simple/elegant to me but I'll gladly revert it if you want. 

- regarding balanced spacing in params, I tried to go for the rule that enforced the least changes. But usually the related eslint-rules are [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing) and [array-bracket-spacing](http://eslint.org/docs/rules/array-bracket-spacing), so one issue with the code was that import statements did not respect curly-spacing ever, and it was mixed for inline objects.

I'll see if I can cherrypick the eslint commit to another PR.

**Comment by alexanderGugel** [comment]
@mgcrea It looks like this conflicted with the dependency update PR. Can you rebase so we can merge? Really like this.

File: src/debuglog.js (hunk 5 above)
Comment (alexanderGugel) [inline src/debuglog.js]: Ahhh I see. Ok, well, let's keep it then. The linting rule definitely makes sense.

**Comment by mgcrea** [comment]
@alexanderGugel, I'm on it.

**Comment by mgcrea** [comment]
Travis is green, ready for merge! Thanks for the review.

**Pull request merged**